pkg/controller/installation/merge.go

```diff
@@ -0,0 +1,189 @@
+
+	switch compareFields(inst.Spec.Variant, override.Spec.Variant) {
+	case BOnlySet, Different:
+		inst.Spec.Variant = override.Spec.Variant
+	}
+
+	switch compareFields(inst.Spec.Registry, override.Spec.Registry) {
+	case BOnlySet, Different:
+		inst.Spec.Registry = override.Spec.Registry
+	}
+
+	switch compareFields(inst.Spec.ImagePath, override.Spec.ImagePath) {
+	case BOnlySet, Different:
+		inst.Spec.ImagePath = override.Spec.ImagePath
+	}
+
+	switch compareFields(inst.Spec.ImagePullSecrets, override.Spec.ImagePullSecrets) {
+	case BOnlySet, Different:
+		inst.Spec.ImagePullSecrets = make([]v1.LocalObjectReference, len(override.Spec.ImagePullSecrets))
+		copy(inst.Spec.ImagePullSecrets, override.Spec.ImagePullSecrets)
+	}
+
+	switch compareFields(inst.Spec.KubernetesProvider, override.Spec.KubernetesProvider) {
+	case BOnlySet, Different:
+		inst.Spec.KubernetesProvider = override.Spec.KubernetesProvider
+	}
+
+	switch compareFields(inst.Spec.CNI, override.Spec.CNI) {
+	case BOnlySet:
+		inst.Spec.CNI = override.Spec.CNI.DeepCopy()
+	case Different:
+		inst.Spec.CNI = mergeCNISpecs(inst.Spec.CNI, override.Spec.CNI)
+	}
+
+	switch compareFields(inst.Spec.CalicoNetwork, override.Spec.CalicoNetwork) {
+	case BOnlySet:
+		inst.Spec.CalicoNetwork = override.Spec.CalicoNetwork.DeepCopy()
+	case Different:
+		inst.Spec.CalicoNetwork = mergeCalicoNetwork(inst.Spec.CalicoNetwork, override.Spec.CalicoNetwork)
+	}
+
+	switch compareFields(inst.Spec.ControlPlaneNodeSelector, override.Spec.ControlPlaneNodeSelector) {
+	case BOnlySet, Different:
+		inst.Spec.ControlPlaneNodeSelector = make(map[string]string, len(override.Spec.ControlPlaneNodeSelector))
+		for key, val := range override.Spec.ControlPlaneNodeSelector {
+			inst.Spec.ControlPlaneNodeSelector[key] = val
+		}
+	}
+
+	switch compareFields(inst.Spec.NodeMetricsPort, override.Spec.NodeMetricsPort) {
+	case BOnlySet, Different:
+		inst.Spec.NodeMetricsPort = override.Spec.NodeMetricsPort
+	}
+
+	switch compareFields(inst.Spec.FlexVolumePath, override.Spec.FlexVolumePath) {
+	case BOnlySet, Different:
+		inst.Spec.FlexVolumePath = override.Spec.FlexVolumePath
+	}
+
+	switch compareFields(inst.Spec.NodeUpdateStrategy, override.Spec.NodeUpdateStrategy) {
+	case BOnlySet, Different:
+		override.Spec.NodeUpdateStrategy.DeepCopyInto(&inst.Spec.NodeUpdateStrategy)
+	}
+
+	switch compareFields(inst.Spec.ComponentResources, override.Spec.ComponentResources) {
+	case BOnlySet, Different:
+		inst.Spec.ComponentResources = make([]operatorv1.ComponentResource, len(override.Spec.ComponentResources))
+		copy(inst.Spec.ComponentResources, override.Spec.ComponentResources)
+	}
+
+	return inst
+}
+
+func compareFields(a, b interface{}) CompareResult {
+	az := reflect.DeepEqual(a, reflect.Zero(reflect.TypeOf(a)).Interface())
+	bz := reflect.DeepEqual(b, reflect.Zero(reflect.TypeOf(b)).Interface())
+	if az && bz {
+		return Same
+	}
+	if reflect.DeepEqual(a, b) {
+		return Same
+	}
+	if az {
+		return BOnlySet
```

Comment:
I might not be reading this code correctly, but it looks like when  az and bz are both set to different values, we incorrectly return `BOnlySet`

Reply:
az and bz will be true if the associated a or b is the nil or zero value, otherwise it will be false.
If they both are the zero value then they're the same and returned after the 1st check.
If only one is set then that should cause the appropriate BOnlySet or AOnlySet to be returned.
If both are set then nither `if az` or `if bz` will hit.

I'm adding a comment to the `az` and `bz` setting to make it clearer what they do.